### PR TITLE
[ECF-1-391] Add paper-trail auditing

### PR DESCRIPTION
**Pull request by amazimbe**
### Context
As... DfE
I need to know what actions have been performed on the system by which DfE users
So that I have an audit trail if something goes wrong and know who to contact if I have a question

### Testing
Acceptance criteria

- It's done when we have a record of all create/edit/delete actions for School Induction Co-ordinators 
- It's done when we have a record of all create/edit/delete actions for Delivery Partners
- It's done when we have a record of all create/edit/delete actions for Lead Provider USERS (not organisations)
- It's done when we have a record of all create/amend/de-register actions for partnerships between Lead Providers and Schools

### CODE REVIEW

File: app/controllers/admin/base_controller.rb [hunk 1]
```diff
@@ -3,6 +3,7 @@
 class Admin::BaseController < ApplicationController
   include Pundit
 
+  before_action :set_paper_trail_whodunnit
```
Comment (kicferk1) [inline app/controllers/admin/base_controller.rb:6]: Should this be the first action? What happens someone tries to view the page without logging in?

File: db/migrate/20210319153015_create_versions.rb [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+  # when serializing very large objects, `text` might not be big enough.
+  TEXT_BYTES = 1_073_741_823
+
+  def change
+    create_table :versions do |t|
+      t.string   :item_type, { null: false }
+      t.bigint   :item_id,   null: false
+      t.string   :event,     null: false
+      t.string   :whodunnit
+      t.text     :object, limit: TEXT_BYTES
```
Comment (kicferk1) [inline db/migrate/20210319153015_create_versions.rb:18]: So does it store a serialised object? Json? Something else?

File: db/migrate/20210319153015_create_versions.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      # MySQL timestamp columns do not support fractional seconds unless
+      # defined with "fractional seconds precision". MySQL users should manually
+      # add fractional seconds precision to this migration, specifically, to
+      # the `created_at` column.
+      # (https://dev.mysql.com/doc/refman/5.6/en/fractional-seconds.html)
+      #
+      # MySQL users should also upgrade to at least rails 4.2, which is the first
+      # version of ActiveRecord with support for fractional seconds in MySQL.
+      # (https://github.com/rails/rails/pull/14359)
+      #
```
Comment (kicferk1) [inline db/migrate/20210319153015_create_versions.rb:32]: Do we care about mysql comment? We use postgres after all.

File: db/migrate/20210319153015_create_versions.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
+# frozen_string_literal: true
+
+# This migration creates the `versions` table, the only schema PT requires.
+# All other migrations PT provides are optional.
+class CreateVersions < ActiveRecord::Migration[6.1]
+  # The largest text column available in all supported RDBMS is
+  # 1024^3 - 1 bytes, roughly one gibibyte.  We specify a size
+  # so that MySQL will use `longtext` instead of `text`.  Otherwise,
+  # when serializing very large objects, `text` might not be big enough.
```
Comment (kicferk1) [inline db/migrate/20210319153015_create_versions.rb:9]: Ditto, I wonder if we need this in postgres.

I also wonder how large we expect our objects to get - I would be shocked if they were longer than 10k characters.

File: db/schema.rb [hunk 5]
```diff
@@ -171,6 +171,10 @@
     t.uuid "school_id", null: false
     t.uuid "lead_provider_id", null: false
     t.uuid "cohort_id", null: false
+    t.string "status", default: "pending", null: false
+    t.string "reason_for_rejection"
```
Comment (ethanmills) [inline db/schema.rb:175]: Think you need to regenerate the schema?

File: app/controllers/admin/base_controller.rb (hunk 1 above)
Comment (amazimbe) [reply app/controllers/admin/base_controller.rb]: I've moved it to the end

File: db/migrate/20210319153015_create_versions.rb (hunk 2 above)
Comment (amazimbe) [reply db/migrate/20210319153015_create_versions.rb]: I've changed the schema to store this as json

File: db/migrate/20210319153015_create_versions.rb (hunk 3 above)
Comment (amazimbe) [reply db/migrate/20210319153015_create_versions.rb]: The comment came with the auto generator. I've removed the mysql comment

File: spec/requests/admin/administrators/administrators_spec.rb [hunk 6]
```diff
@@ -167,6 +167,16 @@
       expect(admin_profile.discarded?).to be false
       expect(admin_user.discarded?).to be false
     end
+
+    describe "when an audited action", versioning: true do
```
Comment (amazimbe) [inline spec/requests/admin/administrators/administrators_spec.rb:171]: I added a few of these tests for sanity checking and mostly to confirm whodunnit is being set to the current user.

File: db/migrate/20210319153015_create_versions.rb (hunk 4 above)
Comment (amazimbe) [reply db/migrate/20210319153015_create_versions.rb]: I've removed it and moved to json rather than text.

File: db/schema.rb (hunk 5 above)
Comment (amazimbe) [reply db/schema.rb]: 👍 That's now been sorted

**Comment by amazimbe** [comment]
[Implement audit trail for DfE Admin journey](https://trello.com/c/uAiEFAX8/391-implement-audit-trail-for-dfe-admin-journey)

File: app/models/delivery_partner.rb [hunk 7]
```diff
@@ -4,6 +4,8 @@ class DeliveryPartner < DiscardableRecord
   has_many :provider_relationships
   has_many :lead_providers, through: :provider_relationships
 
+  has_paper_trail
```
Comment (kicferk1) [inline app/models/delivery_partner.rb:7]: I wonder if we want to form a convention that paper trail is the very first thing in the model, if it has it - so we don't spend time looking for it. Thoughts?

**Review by kicferk1 — APPROVED** [review]
Looks good.

File: app/models/delivery_partner.rb (hunk 7 above)
Comment (amazimbe) [reply app/models/delivery_partner.rb:7]: No strong opinions either way. I'm sure there will be other things we add that we may want to come first and then soon or later we would have to make a choice and I see that convention breaking subsequently.

Comment (kicferk1) [reply app/models/delivery_partner.rb:7]: Yeah, I guess that convention might evolve to "things from external gems come first, then things from Rails, then whatever custom methods we wrote", or something to that effect

**Pull request merged**